blockchain/chainio.go

```diff
@@ -1167,6 +1167,44 @@ func (b *BlockChain) initChainState() error {
 			return err
 		}
 
+		// Load all of the headers from the data for the known best
+		// chain and construct the block index accordingly.  Since the
+		// number of nodes are already known, perform a single alloc
+		// for them versus a whole bunch of little ones to reduce
+		// pressure on the GC.
```

Comment:
A similar memory optimization (reducing allocations) could also be done within the index which maps a blockhash to a `blockNode`.